### DISCUSSION TITLE
Plausible fails to connect to clickhouse even with correct CLICKHOUSE_DATABASE_URL

### DISCUSSION THREAD

**Discussion by xrl** [plausible/analytics#3070] [Self-Hosted Support]
I am spinning up plausible+clickhouse inside of Kubernetes using argo-cd. I have one argo application to manage the plausible helm release (homemade chart), and another argo application to manage the clickhouse helm release (bitnami clickhouse).

Here are the application definitions:

```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: plausible
  namespace: argo
spec:
  project: default
  source:
    repoURL: https://github.com/my-org/plausible.git
    targetRevision: HEAD
    path: charts/plausible
    helm:
      releaseName: plausible
  destination:
    server: https://kubernetes.default.svc
    namespace: plausible
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: plausible-clickhouse
  namespace: argo
spec:
  project: default
  source:
    repoURL: https://charts.bitnami.com/bitnami
    targetRevision: 3.4.4
    chart: clickhouse
    helm:
      releaseName: clickhouse
  destination:
    server: https://kubernetes.default.svc
    namespace: plausible
```

I have confirmed that clickhouse+plausible have happy pods:

```
% k get pods
NAME                          READY   STATUS    RESTARTS   AGE
clickhouse-5b498cb688-fwqfr   1/1     Running   0          6d20h
clickhouse-shard0-0           1/1     Running   0          8h
clickhouse-shard0-1           1/1     Running   0          21h
clickhouse-shard0-2           1/1     Running   0          21h
clickhouse-shard1-0           1/1     Running   0          16h
clickhouse-shard1-1           1/1     Running   0          21h
clickhouse-shard1-2           1/1     Running   0          21h
clickhouse-zookeeper-0        1/1     Running   0          21h
clickhouse-zookeeper-1        1/1     Running   0          16h
clickhouse-zookeeper-2        1/1     Running   0          8h
plausible-8587857b77-pd424    1/1     Running   0          18h
```

And the services are laid out like this:

```
% k get svc 
NAME                            TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                                        AGE
clickhouse                      ClusterIP   10.100.94.202    <none>        8123/TCP,9000/TCP,9004/TCP,9005/TCP,9009/TCP   10d
clickhouse-headless             ClusterIP   None             <none>        8123/TCP,9000/TCP,9004/TCP,9005/TCP,9009/TCP   21h
clickhouse-zookeeper            ClusterIP   10.100.69.11     <none>        2181/TCP,2888/TCP,3888/TCP                     21h
clickhouse-zookeeper-headless   ClusterIP   None             <none>        2181/TCP,2888/TCP,3888/TCP                     21h
plausible                       ClusterIP   10.100.220.100   <none>        8000/TCP                                       9d
```

And plausible is talking to the `clickhouse` service like this:

```
% k get pod plausible-8587857b77-pd424 -o yaml
kind: Pod
metadata:
  name: plausible-8587857b77-pd424
  namespace: plausible
spec:
  containers:
  - env:
    - name: BASE_URL
      value: https://plausible.my-thing.com
    - name: DISABLE_REGISTRATION
      value: "true"
    - name: POSTGRES_PORT
      value: "5432"
    - name: DATABASE_URL
      valueFrom:
        secretKeyRef:
          key: POSTGRES_URL
          name: plausible-postgres-direct-secret
    - name: CLICKHOUSE_DATABASE_URL
      value: http://default:$admin-password-from-the-helm-created-secret@clickhouse:8123/plausible
    envFrom:
    - secretRef:
        name: plausible-postgres-direct-secret
    image: plausible/analytics:v2.0.0-rc.2
    livenessProbe:
      failureThreshold: 3
      httpGet:
        path: /
        port: http
        scheme: HTTP
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 1
    name: plausible
    ports:
    - containerPort: 8000
      name: http
      protocol: TCP
    readinessProbe:
      failureThreshold: 3
      httpGet:
        path: /
        port: http
        scheme: HTTP
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 1
    resources:
      limits:
        cpu: "2"
        memory: 10Gi
      requests:
        cpu: "2"
        memory: 10Gi
```

but when I look at the logs for plausible I get an infinite stream of these lines:

```
01:37:35.801 [error] Ch.Connection (#PID<0.3392.0>) failed to connect: ** (Ch.Error) Code: 81. DB::Exception: Database plausible doesn't exist. (UNKNOWN_DATABASE) (version 23.5.3.24 (official build))
```

but if I query the clickhouse service using curl I can see the database:

```
# echo 'SHOW DATABASES' | curl 'http://default:$admin-password-from-the-helm-created-secret@clickhouse:8123/' --data-binary @-
INFORMATION_SCHEMA
default
information_schema
plausible
system
```

Why is plausible unable to acccess the `plausible` database from clickhouse?

**Comment by ruslandoga** [comment]
👋 @xrl 

Is the `curl` request made from `plausible` container? Can you try a request that queries a table from `plausible` database?

**Reply by xrl** [reply]
No, however, it was made from the same scope DNS and network wise.

**Reply by xrl** [reply]
Someone mentioned I might want a full DNS name since erlang may not respect Kubernete's resolv.conf with ndots and name hierarchy expansion.

**Reply by ruslandoga** [reply]
> No, however, it was made from the same scope DNS and network wise.

It might make things clearer if you try it from the container.

> Someone mentioned I might want a full DNS name since erlang may not respect Kubernete's resolv.conf with ndots and name hierarchy expansion.

It doesn't seem like `plausible` has any problem resolving and connecting to `clickhouse`. That error is a response from ClickHouse.

**Reply by xrl** [reply]
From inside the plausible container I can use `wget`:

```
~ $ wget http://default:yNd6jhJEFN@clickhouse.plausible.svc.cluster.local:8123/plausible
Connecting to clickhouse.plausible.svc.cluster.local:8123 (10.100.94.202:8123)
wget: server returned error: HTTP/1.1 404 Not Found
```

Does the clickhouse URL need to be a valid URL or does plausible parse the URL and remove the path element and use the path/dbname elsewhere?

**Reply by ruslandoga** [reply]
> does plausible parse the URL and remove the path element and use the path/dbname

Yes, it's done by the underlying library, Ecto. The URL is parsed into a structure with `database`, `username`, `password`, `host`, `port`, etc. fields which are then used in our ClickHouse client.

I'm not very familiar with `wget`, but with `curl` you can just put these parameters into the query string:

```console
$ curl 'http://clickhouse:8123/?database=plausible&username=default&password=yNd6jhJEFN' -d 'SELECT count(*) FROM events_v2' -v
```

**Reply by walkerrandolphsmith** [reply]
```
/app $ export CLICKHOUSE_DATABASE_URL="http://plausible-plausible-analytics-events-db:8123/?database=plausible"
/app $ wget --post-data='SELECT 1' -qO- $CLICKHOUSE_DATABASE_URL
1
/app $ export CLICKHOUSE_DATABASE_URL="http://plausible-plausible-analytics-events-db:8123/plausible"
/app $ wget --post-data='SELECT 1' -qO- $CLICKHOUSE_DATABASE_URL
wget: server returned error: HTTP/1.1 404 Not Found
```

**Reply by walkerrandolphsmith** [reply]
However, if I use this format then I get the following error on startup
```
 (EXIT from #PID<0.95.0>) an exception was raised:
    ** (Ecto.InvalidURLError) invalid URL http://user:password@plausible-plausible-analytics-events-db:8123/?database=plausible, path should be a database name. The parsed URL is: %URI{scheme: "http", authority: "user:password@plausible-plausible-analytics-events-db:8123", userinfo: "user:password", host: "plausible-plausible-analytics-events-db", port: 8123, path: "/", query: "database=plausible", fragment: nil}   
        (ecto 3.12.5) lib/ecto/repo/supervisor.ex:111: Ecto.Repo.Supervisor.parse_url/1
        (ecto 3.12.5) lib/ecto/repo/supervisor.ex:29: Ecto.Repo.Supervisor.init_config/4
        (ecto 3.12.5) lib/ecto/repo/supervisor.ex:190: Ecto.Repo.Supervisor.init/1
        (stdlib 6.2.2) supervisor.erl:869: :supervisor.init/1
        (stdlib 6.2.2) gen_server.erl:2229: :gen_server.init_it/2
        (stdlib 6.2.2) gen_server.erl:2184: :gen_server.init_it/6
        (stdlib 6.2.2) proc_lib.erl:329: :proc_lib.init_p_do_apply/3
```

**Reply by ruslandoga** [reply]
👋 @walkerrandolphsmith 

`CLICKHOUSE_DATABASE_URL` format should be like `http://user:password@plausible-plausible-analytics-events-db:8123/plausible`

**Reply by walkerrandolphsmith** [reply]
Sorry I should have been consistent with my previous messages I am using this format http://user:password@plausible-plausible-analytics-events-db:8123/plausible. 

I can use the clickhouse client to confirm the database exists and the specified username and password combination work. I can Also confirm that the plausible app can reach the instance of clickhouse via HTTP using wget --post-data='SELECT 1' -qO- http://user:password@plausible-plausible-analytics-events-db:8123/?database=plausible

**Reply by ruslandoga** [reply]
Would you be able to share `docker compose logs` from a Plausible instance configured with

```
CLICKHOUSE_DATABASE_URL="http://user:password@plausible-plausible-analytics-events-db:8123/plausible"
```

That should give us more information where the problem happens

**Reply by walkerrandolphsmith** [reply]
```
19:50:29.312 [error] GenServer Plausible.Session.WriteBuffer terminating
** (Ch.Error) Code: 516. DB::Exception: : Authentication failed: password is incorrect, or there is no user with such name.

If you have installed ClickHouse and forgot password you can reset it in the configuration file.
The password for default user is typically located at /etc/clickhouse-server/users.d/default-password.xml
and deleting this file will reset the password.
See also /etc/clickhouse-server/users.xml on the server where ClickHouse is installed.

: while executing 'FUNCTION dictHas('plausible.acquisition_channel_paid_sources_dict' :: 47, lower(utm_source) : 15) -> dictHas('plausible.acquisition_channel_paid_sources_dict', lower(utm_source)) UInt8 : 82'. (AUTHENTICATION_FAILED) (version 24.12.6.70 (official build))

    (ecto_sql 3.12.1) lib/ecto/adapters/sql.ex:1096: Ecto.Adapters.SQL.raise_sql_call_error/1
    (stdlib 6.2.2) gen_server.erl:2393: :gen_server.try_terminate/3
    (stdlib 6.2.2) gen_server.erl:2594: :gen_server.terminate/10
    (stdlib 6.2.2) proc_lib.erl:329: :proc_lib.init_p_do_apply/3
Last message: :tick
```

Note that from the resource running clickhouse I get the following

```
plausible-plausible-analytics-events-db-6bf7d6c46f-bf42t:/# clickhouse-client --user user --password 'XXX' --query="SHOW DATABASES"
plausible
```

**Reply by ruslandoga** [reply]
Hm, that's a weird place to get an error :)

I wonder if you could try connecting "without" creds since it seems to have worked in https://github.com/plausible/analytics/discussions/3070#discussioncomment-13712710 (I am assuming those commands were executed from within `plausible` container)

```
CLICKHOUSE_DATABASE_URL="http://plausible-plausible-analytics-events-db:8123/plausible"
```

Other than that, and double checking the password, I don't know what could be wrong ...

---

If nothing works out you could try starting with colocated, no auth, default db setup in https://github.com/plausible/community-edition and building from there, maybe it would be easier to pinpoint the problem then.

**Reply by walkerrandolphsmith** [reply]
```
clickhouse-client --user non_default_user --password='password' --query="SHOW DATABASES"
plausible

clickhouse-client --query="SHOW DATABASES"
plausible
plausible_events_db
...
```

@ruslandoga, Once I use http://plausible-plausible-analytics-events-db:8123/plausible_events_db it works.

**Reply by walkerrandolphsmith** [reply]
I ran GRANT ALL ON plausible_events_db.* TO my_user; and it started working 🥳

**Reply by walkerrandolphsmith** [reply]
@xrl I bet this might be the answer to your original question. Want to try and let us know?

**Comment by kashalls** [comment]
So, bitnami's clickhouse chart for whatever reason wont replicate datatbases from one shard to another. I brought it down to 1 shard and everything worked for whatever reason. I've been trying to get plausible to work with it and I can't even get the container to start right now.

> clickhouse service uses a loadbalancer to all the shards. When plausible connects it picks a shard. The shard the db was created on might not be the same one it reads from and vise versa.

**Reply by ruslandoga** [reply]
@kashalls 👋

Are there any errors logged before the container exists?

**Reply by kashalls** [reply]
@ruslandoga 👋 

None so far, this is a seperate issue but I am having trouble getting plausible to create the database and tables. 
I've been struggling for the past month coming back to this every so often. I can eventually get the container to boot, but it throws Plausble Salt table missing errors.

**Reply by ruslandoga** [reply]
Hm. If you describe your setup a bit more (I guess both of your databases are external?) and share your docker-compose.yml and plausible-conf.env, I could help you troubleshoot :)

**Reply by kashalls** [reply]
Here's my [deployment](https://github.com/kashalls/home-cluster/blob/main/kubernetes/apps/default/plausible/app/helmrelease.yaml) and [env](https://github.com/kashalls/home-cluster/blob/main/kubernetes/apps/default/plausible/app/externalsecret.yaml)

You can find my [clickhouse deployment](https://github.com/kashalls/home-cluster/blob/main/kubernetes/apps/database/clickhouse/app/helmrelease.yaml) there.

As of right now I am dealing with: `sh: /entrypoint.sh db createdb && /entrypoint.sh db migrate && /entrypoint.sh run: not found`. Trying to get plausible to make the tables required.

**Reply by kashalls** [reply]
So pretty much the database is created, but there are no tables. When I try to do the migration, it exists with code 0. Run still fails. Not sure what to do. Explored the container and tried everything I can think of.

**Reply by kashalls** [reply]
@ruslandoga Just mentioning you to see if you have any clues. Can move this to another discussion if needed.

**Reply by ruslandoga** [reply]
@kashalls judging by the error you posted,

> sh: /entrypoint.sh db createdb && /entrypoint.sh db migrate && /entrypoint.sh run: not found

It seems like the problem might be with the [`command:` rule.](https://github.com/kashalls/home-cluster/blob/8463d37b83482625f8e93e72cd950056372b070a/kubernetes/apps/default/plausible/app/helmrelease.yaml#L39-L45) I see you've tried it multiple ways, but have you tried passing the scripts as args like [here,](https://github.com/plausible/hosting/blob/3e1462eabd510cc9320a6cfed0ae45355d86bb05/kubernetes/plausible.yaml#L48-L52) and also maybe moving it to an init container. Then for the actual plausible container, you wouldn't need the custom command or `sh -c` as `/entrypoint.sh run` [is the default.](https://github.com/plausible/analytics/blob/3c76053c3fcd3d6b30efbfd1159ed275e93d901c/Dockerfile#L77-L79)

**Reply by kashalls** [reply]
```
{"Kernel pid terminated",application_controller,"{application_start_failure,plausible,{{shutdown,{failed_to_start_child,'Elixir.Plausible.Session.Salts',{#{'__exception__' => true,'__struct__' => 'Elixir.Postgrex.Error',connection_id => 1256896,message => nil,postgres => #{code => undefined_table,file => <<\"parse_relation.c\">>,line => <<\"1381\">>,message => <<\"relation \\"salts\\" does not exist\">>,pg_code => <<\"42P01\">>,position => <<\"13\">>,routine => <<\"parserOpenTable\">>,severity => <<\"ERROR\">>,unknown => <<\"ERROR\">>},query => <<\"DELETE FROM \\"salts\\" AS s0 WHERE (s0.\\"inserted_at\\" < now() - '48 hours'::interval)\">>},[{'Elixir.Ecto.Adapters.SQL',raise_sql_call_error,1,[{file,\"lib/ecto/adapters/sql.ex\"},{line,913},{error_info,#{module => 'Elixir.Exception'}}]},{'Elixir.Ecto.Adapters.SQL',execute,6,[{file,\"lib/ecto/adapters/sql.ex\"},{line,828}]},{'Elixir.Plausible.Session.Salts','-start_link/1-fun-0-',0,[{file,\"lib/plausible/session/salts.ex\"},{line,8}]},{'Elixir.Agent.Server',init,1,[{file,\"lib/agent/server.ex\"},{line,8}]},{gen_server,init_it,2,[{file,\"gen_server.erl\"},{line,851}]},{gen_server,init_it,6,[{file,\"gen_server.erl\"},{line,814}]},{proc_lib,init_p_do_apply,3,[{file,\"proc_lib.erl\"},{line,240}]}]}}},{'Elixir.Plausible.Application',start,[normal,[]]}}}"}
Kernel pid terminated (application_controller) ({application_start_failure,plausible,{{shutdown,{failed_to_start_child,'Elixir.Plausible.Session.Salts',{#{'__exception__' => true,'__struct__' => 'Elixir.Postgrex.Error',connection_id => 1256896,message => nil,postgres => #{code => undefined_table,file => <<"parse_relation.c">>,line => <<"1381">>,message => <<"relation \"salts\" does not exist">>,pg_code => <<"42P01">>,position => <<"13">>,routine => <<"parserOpenTable">>,severity => <<"ERROR">>,unknown => <<"ERROR">>},query => <<"DELETE FROM \"salts\" AS s0 WHERE (s0.\"inserted_at\" < now() - '48 hours'::interval)">>},[{'Elixir.Ecto.Adapters.SQL',raise_sql_call_error,1,[{file,"lib/ecto/adapters/sql.ex"},{line,913},{error_info,#{module => 'Elixir.Exception'}}]},{'Elixir.Ecto.Adapters.SQL',execute,6,[{file,"lib/ecto/adapters/sql.ex"},{line,828}]},{'Elixir.Plausible.Session.Salts','-start_link/1-fun-0-',0,[{file,"lib/plausible/session/salts.ex"},{line,8}]},{'Elixir.Agent.Server',init,1,[{file,"lib/agent/server.e

Crash dump is being written to: erl_crash.dump...done
```

That's what I am seeing when I do this.

**Reply by ruslandoga** [reply]
When removing `command` rule or using `args:`? The error means that the `salts` table is missing in PostgreSQL. `salts` table is created with `/entrypoint.sh db migrate`

**Reply by kashalls** [reply]
Right. I did the same thing recommended above for both: https://github.com/kashalls/home-cluster/blob/main/kubernetes/apps/default/plausible/app/helmrelease.yaml#L38

**Reply by ruslandoga** [reply]
Can please you check the exit status and logs of the init container? It seems to have failed if there is no `salts` table when `plausible` container starts up. Seems like `BASE_URL` env var might be missing from the init container.

**Reply by kashalls** [reply]
Added that env. the initContainer is showing as exit code 0 and is completed. Does not seem like that command is wanting to create/migrate the database.

**Reply by ruslandoga** [reply]
Have you tried `args` written like this?

```yml
args:
  - /entrypoint.sh db createdb && /entrypoint.sh db migrate
```

I don't know much about YAML, but the last working configuration for k8s had this exact formatting, so it might be worth trying out.

<details><summary>Here's what GPT wrote about that difference</summary>

Yes, there are some key differences between the two YAML snippets.

1. **Structure and Parsing**:
   - In the first snippet:
     ```
     command: ['/bin/sh', '-c']
     args: ['/entrypoint.sh', 'db', 'createdb', '&&', '/entrypoint.sh', 'db', 'migrate']
     ```
     Both `command` and `args` have lists of individual arguments. The shell command (`/bin/sh -c`) expects a single string argument that represents the command to be run. In the provided format, it's not receiving a full single string command. 

   - In the second snippet:
     ```
     command:
       - "/bin/sh"
       - "-c"
     args:
       - /entrypoint.sh db createdb && /entrypoint.sh db migrate
     ```
     The structure is more appropriate for the shell command. The `args` is correctly giving a single string to the shell command as an argument.

2. **Execution Implications**:
   - In the first snippet:
     The way `args` is structured suggests that each item is a separate argument. It means that if passed to a command, the shell will try to execute `/entrypoint.sh` with the following arguments: `db`, `createdb`, `&&`, `/entrypoint.sh`, `db`, `migrate`. The shell won't treat `&&` as a control operator, but rather as an argument. This would likely cause an error because the shell doesn't interpret `&&` correctly.

   - In the second snippet:
     The command `args` is correctly structured as a single string: `/entrypoint.sh db createdb && /entrypoint.sh db migrate`, which the shell (`/bin/sh -c`) can interpret and execute correctly.

In summary, the second YAML snippet is correct if you aim to execute `/entrypoint.sh db createdb` followed by `/entrypoint.sh db migrate`. The first one will likely result in an error due to the incorrect interpretation of the `args`.

</details>

**Reply by kashalls** [reply]
That was it. Its running now,
